### DISCUSSION TITLE
How can I mark the DAG as Failed?

### DISCUSSION THREAD

**Discussion by hbc-acai** [apache/airflow#34382] [Q&A] [answered]
Hello community, 

This might be a very basic question. Is there recommended way to mark a DAG run as failed? For example, an argument passing to EmptyOperator would work for me. My use case is this, consider the following DAG. 

```
    start >> virtualenv_task # read from DB
    is_timeout_task = is_timeout(virtualenv_task) # this is branch operator 
    is_timeout_task >> end # success branch 
    is_timeout_task >> email_error # failure branch
    email_error >> end_with_failure # ??? mark DAG as failed
```

When timeout happens, I want to send an error email and mark the DAG as red. What Operator should I used for `end_with_failure`? 

(One other way is to make the `email_error` task to throw an exception after the email is sent out. Then it would be hard to tell whether the error is from the timeout or another exception happened during the email process. )

**Comment by Taragolis** [comment]
If this all flow then maybe you need to just use callback/notifier here?
- [Callback Types](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/callbacks.html#callback-types)
- [Creating a notifier](https://airflow.apache.org/docs/apache-airflow/stable/howto/notifications.html)

And turn your DAG to something like 
```python
start >> python_virtualen_task >> end  # python_virtualen_task has `on_failure_callback` 
```

**Reply by hbc-acai** [reply]
Thank you. My `python_virtualenv_task` reads from a DB, and I need to measure the actual query time. I only want to send an email alert when the query timeouts. 

- Using `on_failure_callback` is not ideal, because I could not tell whether the failures came from the DB query timeout or other errors. 
- Also, using 'on_sla_callback' might not work ( though I haven't tried), because it looks at the the Task execution time which seems include the time spent on installing the python requirements which can be slow.

I am introducing Airflow to my team so I want to setup some rigorous working examples.

**Reply by potiuk** [reply]
You can write your custom operator to do whatever you want. Writing your own operator is first-class-citizen in airflow. Smth like that:

```
class MyFailingOperator(BaseOperator):
     def execute():
          raise Exception()
```

or even:

```
@task
def fail():
     raise Exception()
```

**Reply by Taragolis** [reply, marked as answer]
> Also, using 'on_sla_callback' might not work ( though I haven't tried), because it looks at the the Task execution time which seems include the time spent on installing the python requirements which can be slow.

I'm not sure that current implementation of SLA would help, there is plan to refactor this feature: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-57+Refactor+SLA+Feature

In general Jarek's solution more straight forward and should be used in all simple cases. I used some custom operator as last step in complex DAGs or on DAG's leaf which check upstream task and failed with appropriate message.

<details>
  <summary>It's pretty old, still work on `main` branch and ⚠️ might use something which not a part of Airflow Public Interface ⚠️</summary>

```python
from collections import deque
from typing import Dict, Iterable, Optional, Set

from airflow.exceptions import AirflowException
from airflow.models import BaseOperator
from airflow.models.taskinstance import TaskInstance
from airflow.sensors.base import BaseSensorOperator
from airflow.utils.state import State


class DagRunTaskFailed(AirflowException):
    """Dag Task Failed Exception

    Raised in case if at least one task failed for DagRun
    """
    ...


class CheckUpstreamSensor(BaseSensorOperator):
    """Waits for complete upstream task in current DagRun
    and check their states.

    Raise an error ``DagRunTaskFailed`` in case if at least one task failed

    :param allowed_states: Iterable of allowed states,
        default is ``['success', 'skipped', 'removed']``
    :type allowed_states: Iterable

    :param failed_states: Iterable of failed or dis-allowed states,
        default is ``['failed', 'upstream_failed']``
    :type failed_states: Iterable

    :param wait_for_completion: Set to `True` to wait until all tasks in not
        ``allowed_states`` and ``failed_states`` otherwise raise on first ``failed_states``
    :type check_existence: bool
    """

    def __init__(
        self,
        allowed_states: Optional[Iterable[str]] = None,
        failed_states: Optional[Iterable[str]] = None,
        wait_for_completion: bool = True,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.failed_states = failed_states or [
            State.FAILED,
            State.UPSTREAM_FAILED,
        ]
        self.allowed_states = allowed_states or [
            State.SUCCESS,
            State.SKIPPED,
            State.REMOVED,
        ]

        if set(self.failed_states).intersection(set(self.allowed_states)):
            raise AirflowException(
                f"Duplicate values provided as allowed `{self.allowed_states}` "
                f"and failed states `{self.failed_states}`"
            )
        self.wait_for_completion = wait_for_completion

    def get_ancestor_task_ids(self) -> Set[str]:
        ancestor_tasks: Set[BaseOperator] = set()
        tasks = deque()

        for t in self.upstream_list:
            tasks.append(t)

        while deque:
            try:
                task = tasks.popleft()
            except IndexError:
                break

            ancestor_tasks.add(task)
            for t in task.upstream_list:
                tasks.append(t)

        return set([t.task_id for t in ancestor_tasks])

    def poke(self, context: Dict) -> bool:
        task_ids = self.get_ancestor_task_ids()
        total_upstream_tasks = len(task_ids)
        if total_upstream_tasks == 0:
            raise AirflowException(f"No upstream tasks for {self.task_id!r}.")

        failed_task_ids = []

        for ti in context["dag_run"].get_task_instances():
            ti: TaskInstance
            task_id = ti.task_id

            if task_id not in task_ids:
                continue

            task_state = ti.state
            if task_state in self.failed_states:
                if not self.wait_for_completion:
                    raise DagRunTaskFailed(f"{task_id} in failed state: {task_state}")

                failed_task_ids.append(task_id)
                task_ids.remove(task_id)
            elif task_state in self.allowed_states:
                task_ids.remove(task_id)

        self.log.info(
            f"Total Upstream task(s) %s. Remain %s. Failed %s",
            total_upstream_tasks,
            len(task_ids),
            len(failed_task_ids)
        )

        if task_ids:  # Upstream task(s) still executed
            return False

        if failed_task_ids:
            raise DagRunTaskFailed(
                f"Selected task_ids in failed states: {', '.join(map(repr, failed_task_ids))}."
            )

        return True
```
  
</details>

In normal circumstances DAG marked Failed in case of all task completed and at least one leaf are Failed, so you just need to be need to be creative and make one of DAG's leaf failed

**Reply by hbc-acai** [reply]
Thank you both. I will use `@task` decorator Jarek proposed. It is nice that Airflow is very flexible.
